File: sdks/python/settlekit/resources/dunning.py
"""Dunning resource — recover failed subscription payments.

Start a campaign for a subscription, record attempt outcomes, and recover (or
let the schedule exhaust).
"""

from __future__ import annotations

from typing import Any, Optional

from ._base import AsyncResource, SyncResource, clean_body, clean_params


class Dunning(SyncResource):
    """Synchronous dunning namespace."""

    def list(self, *, due: bool = False) -> Any:
        """List active campaigns (or only those due). ``GET /v1/dunning``."""
        return self._request(
            "GET", "/v1/dunning", params=clean_params({"due": "true" if due else None})
        )

    def start(self, subscription_id: str) -> Any:
        """Start a dunning campaign for a subscription. ``POST /v1/dunning``."""
        return self._request(
            "POST", "/v1/dunning", body={"subscriptionId": subscription_id}
        )

    def attempt(
        self, subscription_id: str, *, outcome: str, failure_reason: Optional[str] = None
    ) -> Any:
        """Record an attempt outcome (``recovered`` / ``failed``).

        ``POST /v1/dunning/:subscriptionId/attempt``.
        """
        return self._request(
            "POST",
            f"/v1/dunning/{subscription_id}/attempt",
            body=clean_body({"outcome": outcome, "failureReason": failure_reason}),
        )

    def recover(self, subscription_id: str) -> Any:
        """Mark a campaign as recovered. ``POST /v1/dunning/:subscriptionId/recover``."""
        return self._request("POST", f"/v1/dunning/{subscription_id}/recover")


class AsyncDunning(AsyncResource):
    """Asynchronous dunning namespace."""

    async def list(self, *, due: bool = False) -> Any:
        """List active campaigns (or only those due). ``GET /v1/dunning``."""
        return await self._request(
            "GET", "/v1/dunning", params=clean_params({"due": "true" if due else None})
        )

    async def start(self, subscription_id: str) -> Any:
        """Start a dunning campaign for a subscription. ``POST /v1/dunning``."""
        return await self._request(
            "POST", "/v1/dunning", body={"subscriptionId": subscription_id}
        )

    async def attempt(
        self, subscription_id: str, *, outcome: str, failure_reason: Optional[str] = None
    ) -> Any:
        """Record an attempt outcome (``recovered`` / ``failed``).

        ``POST /v1/dunning/:subscriptionId/attempt``.
        """
        return await self._request(
            "POST",
            f"/v1/dunning/{subscription_id}/attempt",
            body=clean_body({"outcome": outcome, "failureReason": failure_reason}),
        )

    async def recover(self, subscription_id: str) -> Any:
        """Mark a campaign as recovered. ``POST /v1/dunning/:subscriptionId/recover``."""
        return await self._request("POST", f"/v1/dunning/{subscription_id}/recover")

File: sdks/rust/src/resources/dunning.rs
//! Dunning resource (`/v1/dunning`).
//!
//! Dunning recovers failed subscription payments: start a campaign for a
//! subscription, record attempt outcomes, and recover (or let it exhaust).

use reqwest::Method;
use serde::Serialize;

use crate::{Client, DunningState, Result};

/// Accessor for dunning operations.
#[derive(Debug)]
pub struct Dunning<'a> {
    client: &'a Client,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct StartBody<'r> {
    subscription_id: &'r str,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct AttemptBody<'r> {
    outcome: &'r str,
    #[serde(skip_serializing_if = "Option::is_none")]
    failure_reason: Option<&'r str>,
}

impl<'a> Dunning<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Dunning { client }
    }

    /// List active dunning campaigns, or only those due when `due_only`.
    pub async fn list(&self, due_only: bool) -> Result<Vec<DunningState>> {
        let path = if due_only {
            "/v1/dunning?due=true"
        } else {
            "/v1/dunning"
        };
        self.client.request_no_body(Method::GET, path).await
    }

    /// Start a dunning campaign for a subscription with a failed payment.
    pub async fn start(&self, subscription_id: &str) -> Result<DunningState> {
        let body = StartBody { subscription_id };
        self.client
            .request(Method::POST, "/v1/dunning", Some(&body))
            .await
    }

    /// Record an attempt outcome (`recovered` / `failed`). `recovered` closes
    /// the campaign; `failed` advances or exhausts it.
    pub async fn attempt(
        &self,
        subscription_id: &str,
        outcome: &str,
        failure_reason: Option<&str>,
    ) -> Result<DunningState> {
        let body = AttemptBody {
            outcome,
            failure_reason,
        };
        self.client
            .request(
                Method::POST,
                &format!("/v1/dunning/{subscription_id}/attempt"),
                Some(&body),
            )
            .await
    }

    /// Mark a subscription's dunning campaign as recovered.
    pub async fn recover(&self, subscription_id: &str) -> Result<DunningState> {
        self.client
            .request::<_, ()>(
                Method::POST,
                &format!("/v1/dunning/{subscription_id}/recover"),
                None,
            )
            .await
    }
}
